### PR TITLE
Make ImmutableValidator more precise while analyzing branches

### PR DESCRIPTION
**Pull request by StrongerXi**
## Motivation

Addresses #12864

## Mechanism

1. Delay error generation. Originally we issue error as soon as we see any write to immutable state variable inside a branch, that's can be too conservative, e.g., when both branches write to the variable.
2. Collect all immutable state variable written in a branch (and their position, obtained via the write expression). After analyzing both branch, we can inspect the set difference, and determine which ones are not written (thus initialized) in both paths. Then we issue error for those variables, _and_ collect the all written to continue this process while avoiding duplicated error generation.

## Backward Compatibility

This shouldn't break backward compatibility because we are generating less errors, i.e., correct programs remain correct. The error message is changed, but we can discuss and tweak that.

## Test Case

```solidity
contract Immutable {
    uint256 public immutable iamImmutable;

    constructor(bool shouldInitWithZero) {
        if (shouldInitWithZero) {
            if (shouldInitWithZero) {
                iamImmutable = 0;
            } else {
                iamImmutable = 1;
            }
        } else {
            if (true) {
                iamImmutable = 1;
            }
        }
    }
}
```
This generates a single error at the last write to `iamImmutable`.
If we remove the last conditional `if (true)`, no error is issued.

## Alternative

I actually think the Java way of handling such errors is worth to consider, i.e., it's no less precise than what this PR aims to achieve, and it issues a single error message at the variable declaration site, saying `the final field ... may not have been initialized`. Which is similar to error 2658 in Solidity.

This approach makes the algorithm a bit simpler -- we just need to keep track and filter out the variables-not-initialized-in-both-branches when we issue error 2658, but we lose the location of the problematic write.

### CODE REVIEW

**Comment by StrongerXi** [comment]
The more I think about it, the more I believe the alternative approach is preferable. Especially true for cases like this:
```solidity
contract Immutable {
    uint256 public immutable iamImmutable;

    constructor(bool shouldInitWithZero) {
        if (shouldInitWithZero) {
            if (shouldInitWithZero) {
                iamImmutable = 0;
            }
        } else {
            if (true) {
                iamImmutable = 1;
            }
        }
    }
}
```

**Comment by chriseth** [comment]
Thank you for the pull request! I fear that the complete solution is a bit more complicated. Can you just ignore all the cases after the `else if (m_inBranch)`?

**Comment by StrongerXi** [comment]
> I fear that the complete solution is a bit more complicated

Would you mind elaborating on what edge case I might be missing? Based on my limited understanding of Solidity, it seems sound if we take the alternative approach, i.e., mark any variable written only in 1 of the if/else branches to be "not properly initialized".

> Can you just ignore all the cases after the else if (m_inBranch)?

I might be misunderstanding, do you mean deleting all these chunks?
```c++
		else if (m_initializedStateVariables.count(&_variableReference))
		{
			if (!read)
				m_errorReporter.typeError(
					1574_error,
					_expression.location(),
					"Immutable state variable already initialized."
				);
			else
				m_errorReporter.typeError(
					2718_error,
					_expression.location(),
					"Immutable variables cannot be modified after initialization."
				);
		}
		else if (read)
			m_errorReporter.typeError(
				3969_error,
				_expression.location(),
				"Immutable variables must be initialized using an assignment."
			);
		m_initializedStateVariables.emplace(&_variableReference)
```

**Comment by StrongerXi** [comment]
Rebase.

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (Marenz) [inline libsolidity/analysis/ImmutableValidator.cpp:109]: It seems dangerous to me to use `std::move()` here. You also access the moved variable a a few lines down and then move it again, yet to my understanding variables that have been moved should no longer be used...

Comment (Marenz) [inline libsolidity/analysis/ImmutableValidator.cpp]: It seems dangerous to me to use std::move() here. You also access the moved variable a a few lines down and then move it again, yet to my understanding variables that have been moved should no longer be used...

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (Marenz) [inline libsolidity/analysis/ImmutableValidator.cpp]: Why the change away from `bool`?

**Review by Marenz — CHANGES_REQUESTED** [review]

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (StrongerXi) [reply libsolidity/analysis/ImmutableValidator.cpp]: Good point. Reverted it to keep changes minimal

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (StrongerXi) [reply libsolidity/analysis/ImmutableValidator.cpp]: I thought about that at first, then [this post](https://stackoverflow.com/questions/7027523/what-can-i-do-with-a-moved-from-object) convinced me that it's safe to use `clear` on it, which would put it back into a valid state.

The motivation behind using a `std::move` here is to avoid unnecessary copying. It'll be a no-issue if we use an immutable map here.

Regardless, I'm still waiting to hear feedback on the alternative approach, since it'll lead to simpler code and better error message IMO.

**Comment by Marenz** [comment]
I am not sure what exactly your suggestion for the alternative approach is.

> This approach makes the algorithm a bit simpler -- we just need to keep track and filter out the variables-not-initialized-in-both-branches when we issue error 2658, but we lose the location of the problematic write.

Problematic write means a conditional write?

Could you explain the alternative approach a bit?

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (Marenz) [reply libsolidity/analysis/ImmutableValidator.cpp]: True, with the info in your link, the usage of `std::move` here seems okay.

**Comment by chriseth** [comment]
The main issue I have with this PR is that I'm not convinced that it properly detects multiple writes to a variable in the same branch. Reporting the error about "not initialized" on the variable seems fine to me.

**Comment by chriseth** [comment]
The main issue I have with this PR is that I'm not convinced that it properly detects multiple writes to a variable in the same branch. Reporting the error about "not initialized" on the variable seems fine to me.

**Comment by StrongerXi** [comment]
@chriseth @Marenz I updated using the alternative approach. The extra check in `ImmutableValidator::checkAllVariablesInitialized` should make the high-level logic pretty clear.

Regarding proper detection of "multiple writes to a variable in the same branch", I believe error 2718 already takes care of it. I also added a test case to illustrate that.

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (chriseth) [inline libsolidity/analysis/ImmutableValidator.cpp]: I think if this is not always an error case, we cannot continue with "else if".

Can you create a test case that detects this bug?

Comment (StrongerXi) [reply libsolidity/analysis/ImmutableValidator.cpp]: Thanks for catching this; I think I have a better understanding of your earlier concerns now.

This PR essentially (1) relaxes the multiple write check to allow single write in each branch, and (2) checks for at least a single write in each branch. I made necessary changes for (1) and augmented test case to demonstrate that it falls back to other existing checks in relaxed cases.

**Comment by StrongerXi** [comment]
Rebase.

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (Marenz) [inline libsolidity/analysis/ImmutableValidator.cpp]: We're using this set template at least three times. I would suggest to make an alias for it and use it  here and in the class declaration..

Comment (Marenz) [reply libsolidity/analysis/ImmutableValidator.cpp]: if we have an alias we can also use it for `prevWrites` instead of `auto`

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (Marenz) [inline libsolidity/analysis/ImmutableValidator.cpp]: Please keep using the style with `{` on a new line

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (Marenz) [inline libsolidity/analysis/ImmutableValidator.cpp]: same here, 
```suggestion
		}
		else
		{
```

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (Marenz) [inline libsolidity/analysis/ImmutableValidator.cpp]:
```suggestion
bool ImmutableValidator::writtenInCurrentControlFlow(VariableDeclaration const& _variableReference) const {
```

This name seems easier to read to me and conveys the same information

File: test/libsolidity/syntaxTests/immutable/conditionally_initialized.sol
Comment (Marenz) [inline test/libsolidity/syntaxTests/immutable/conditionally_initialized.sol]: The test case is comprehensive, but it would be nice to have one test case for each variable-test here, instead of one large one.
Easier to read, faster to see what fails, etc..

**Review by Marenz — CHANGES_REQUESTED** [review]
I didn't find any code logic errors, mostly style.
Oh, a changelog entry is also missing.
Other than that, looks good to me.

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (StrongerXi) [reply libsolidity/analysis/ImmutableValidator.cpp]: Good call, done.

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (StrongerXi) [reply libsolidity/analysis/ImmutableValidator.cpp]: Done.

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (StrongerXi) [reply libsolidity/analysis/ImmutableValidator.cpp]: Done.

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (StrongerXi) [reply libsolidity/analysis/ImmutableValidator.cpp]: Done.

File: test/libsolidity/syntaxTests/immutable/conditionally_initialized.sol
Comment (StrongerXi) [reply test/libsolidity/syntaxTests/immutable/conditionally_initialized.sol]: Agree, done.

**Comment by StrongerXi** [comment]
> I didn't find any code logic errors, mostly style. Oh, a changelog entry is also missing. Other than that, looks good to me.

Thanks!

1. All comments are addressed
2. Changelog entry added
3. `ImmutableValidator.h` documentation updated.

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (Marenz) [inline libsolidity/analysis/ImmutableValidator.cpp]: I didn't catch it earlier, but we use `camelCase` here, so `thenWrites` would be the correct style, same for `elseWrites`

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (Marenz) [inline libsolidity/analysis/ImmutableValidator.cpp]: Style `{`

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (Marenz) [inline libsolidity/analysis/ImmutableValidator.cpp]: `{`

**Review by Marenz — CHANGES_REQUESTED** [review]

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (StrongerXi) [reply libsolidity/analysis/ImmutableValidator.cpp]: Done

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (StrongerXi) [reply libsolidity/analysis/ImmutableValidator.cpp]: Done

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (StrongerXi) [reply libsolidity/analysis/ImmutableValidator.cpp]: Done.

**Comment by StrongerXi** [comment]
1. Addressed all comments.
2. Added comments to help clarify code logic.
3. Rebase.

**Review by Marenz — DISMISSED** [review]

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (chriseth) [inline libsolidity/analysis/ImmutableValidator.cpp:110]: At this point, `m_immutableStateVariablesWrittenInBranch` is in an undefined state. We should clear it an assert it is cleared after the `accept`. Why is it safe to move it here anyway? Can't there be a write operation inside the if condition?

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (chriseth) [inline libsolidity/analysis/ImmutableValidator.cpp]:
```suggestion
	vector<ASTNode::SourceLocation const*> badWriteLocations;
```
(style guide also says "no abbreviations", or at least keep them to a minimum)

Comment (chriseth) [reply libsolidity/analysis/ImmutableValidator.cpp]: Oh it's actually not even used?

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (chriseth) [inline libsolidity/analysis/ImmutableValidator.cpp]:
```suggestion
	for (auto var: thenWrites)
		if (elseWrites.count(var))
			m_initializedStateVariables.emplace(var);
	m_stateVariablesNotInitializedInBothBranches += (elseWrites - thenWrites) + (thenWrites - elseWrites);
```

File: libsolidity/analysis/ImmutableValidator.h
Comment (chriseth) [inline libsolidity/analysis/ImmutableValidator.h]: Wouldn't `m_stateVariablesInitializedInOnlyOneBranch` be a better name?

Comment (chriseth) [reply libsolidity/analysis/ImmutableValidator.h]: (Variables not initialized in any of the branches are not part of the set)

**Comment by chriseth** [comment]
Does this need updated documentation?

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (Marenz) [reply libsolidity/analysis/ImmutableValidator.cpp:110]: That's not strictly true. It is in an `unspecified` state. 

from https://stackoverflow.com/questions/7027523/what-can-i-do-with-a-moved-from-object:

> Objects of types defined in the C++ standard library may be moved from (12.8). Move operations may be explicitly specified or implicitly generated. Unless otherwise specified, such moved-from objects shall be placed in a valid but unspecified state.

The main question now is if `emplace` is an operation that requires preconditions or not. I am still researching that question

Comment (Marenz) [reply libsolidity/analysis/ImmutableValidator.cpp:110]: Or another phrasing from cppreference.com:

> in a "valid but unspecified state", meaning the object's class invariants hold (so functions without preconditions, such as the assignment operator, can be safely used on the object after it was moved from):

I interpret that, that it is completely okay to use `emplace` or `insert` after it was moved. I am not aware of any `valid` state that would prevent insertions in a `std::set` (e.g. you can't do anything to an instance to prevent `emplace` from working except memory exhaustion)

Comment (Marenz) [reply libsolidity/analysis/ImmutableValidator.cpp:110]: That said, having an explicit clear after the move would def. make it failsafe and will avoid the next code reader wondering the same thing.

**Review by leonardoalt — CHANGES_REQUESTED** [review]
I'm revoking the approval so that this doesn't get accidentally merged since there are still reviewer questions/comments.

**Comment by Marenz** [comment]
@StrongerXi would you like to address the open comments or should we take the PR over from here?

**Comment by StrongerXi** [comment]
> @StrongerXi would you like to address the open comments or should we take the PR over from here?

My apologies. Got very busy during the last few weeks. Will push a commit by end of today.

**Comment by Marenz** [comment]
> My apologies. Got very busy during the last few weeks. Will push a commit by end of today.

All good, just pinging to see whats up :)

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (StrongerXi) [reply libsolidity/analysis/ImmutableValidator.cpp:110]: Added explicit `clear` for applicable cases.

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (StrongerXi) [reply libsolidity/analysis/ImmutableValidator.cpp]: My bad -- left from old commit. Removed.

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (StrongerXi) [reply libsolidity/analysis/ImmutableValidator.cpp]: Done -- also added `m_initializedStateVariables.emplace(var);` in the if branch.

**Comment by StrongerXi** [comment]
> Does this need updated documentation?

I already updated the documentation in the header.

File: libsolidity/analysis/ImmutableValidator.h
Comment (StrongerXi) [reply libsolidity/analysis/ImmutableValidator.h]: Good call, done.

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (StrongerXi) [inline libsolidity/analysis/ImmutableValidator.cpp]: Removed `util::contains(m_stateVariablesNotInitializedInBothBranches, varDecl)` -- it's redundant

**Comment by Marenz** [comment]
> I already updated the documentation in the header.

In this case "documentation" means the files under `docs/`, our user-side documentation

**Comment by StrongerXi** [comment]
> > I already updated the documentation in the header.
> 
> In this case "documentation" means the files under `docs/`, our user-side documentation

I see. Just did some `grep`s in `docs/` and the only relevant section seems to be [this](https://github.com/ethereum/solidity/blob/develop/docs/contracts/constant-state-variables.rst#immutable). It doesn't mention anything about initialization constraints tho.

What do you think we should do here?

**Comment by StrongerXi** [comment]
Rebase.

**Review by Marenz — DISMISSED** [review]

File: test/libsolidity/syntaxTests/immutable/conditionally_initialized_multiple_writes.sol
Comment (nikola-matic) [inline test/libsolidity/syntaxTests/immutable/conditionally_initialized_multiple_writes.sol]: Is there a "happy case" for this, i.e. a test like this one here, except without the third write, e.g.

```
contract C {
    uint immutable multipleWrites;

    constructor() {
        if (true) {
            multipleWrites = 1;
        } else {
            multipleWrites = 2;
        }
    }
}
```

Comment (StrongerXi) [reply test/libsolidity/syntaxTests/immutable/conditionally_initialized_multiple_writes.sol]: Good call. Updated all test cases to add such cases.

**Comment by StrongerXi** [comment]
Address comment, rebase, & update changelog.

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (nikola-matic) [inline libsolidity/analysis/ImmutableValidator.cpp]: @chriseth is not a big fan of generous `auto` usage, and whilst I somewhat disagree with him, it's part of our code style, so please unless the type is present in the statement, e.g. `auto const ptr = std::make_shared<Type>(value)`, replace all instances of `auto` with the actual type.

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (nikola-matic) [inline libsolidity/analysis/ImmutableValidator.cpp]: So I know @Marenz had a concern with the usage of `std::move` up top, but in this case I would prefer to simply do an  assignment
```c++
m_immutableStateVariablesWrittenInBranch = prevWrites;
```
as we might add code later on that will access `prevWrites` after the move, causing us to spend time figuring out why we're getting undefined behaviour.

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (nikola-matic) [inline libsolidity/analysis/ImmutableValidator.cpp]: Not sure what the type of `var` is (hence the `auto` comment), but unless it's using perfect forwarding, this is gonna be equivalent to `insert(...)`. This is a nit though, so feel free to ignore this comment even if I'm right.

File: test/libsolidity/syntaxTests/immutable/conditionally_initialized_nested_branch_write.sol
Comment (nikola-matic) [inline test/libsolidity/syntaxTests/immutable/conditionally_initialized_nested_branch_write.sol]: Please move the happy case to a separate file.

**Review by nikola-matic — CHANGES_REQUESTED** [review]
I believe you also need to add a boost test to `test/libsolidity/analysis/FunctionCallGraph.cpp`, but I may be mistaken. See `BOOST_AUTO_TEST_CASE(immutable_initialization)`.

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (StrongerXi) [reply libsolidity/analysis/ImmutableValidator.cpp]: This is a meant to be an efficient "restoration" of the `m_immutableStateVariablesWrittenInBranch` context member variable. So `prevWrites` will not (and should not) be used for any other purposes.

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (StrongerXi) [reply libsolidity/analysis/ImmutableValidator.cpp]: Done

File: libsolidity/analysis/ImmutableValidator.cpp
Comment (StrongerXi) [reply libsolidity/analysis/ImmutableValidator.cpp]: Agree, changed `auto` to explicit type and `emplace` to `insert`.

File: test/libsolidity/syntaxTests/immutable/conditionally_initialized_nested_branch_write.sol
Comment (StrongerXi) [reply test/libsolidity/syntaxTests/immutable/conditionally_initialized_nested_branch_write.sol]: Done.

**Comment by StrongerXi** [comment]
> I believe you also need to add a boost test to `test/libsolidity/analysis/FunctionCallGraph.cpp`, but I may be mistaken. See `BOOST_AUTO_TEST_CASE(immutable_initialization)`.

Thanks, done.

**Comment by StrongerXi** [comment]
Offline chat -- a more general approach is preferred instead of covering control flow one-by-one (e.g., another case would be #13433).

**Review by nikola-matic — COMMENTED** [review]
Hey @StrongerXi, so we've discussed this on our team language design call, and the opinions are somewhat split regarding whether to merge this in, or close it in lieu of a more general solution. Most are fine with merging it in, but since @ekpyron would prefer a general approach, and he's off until October, we'll have to wait for his approval, so if we forget, ping us from October 1st to resume discussion.

On the other hand, we're happy with the contribution, so feel free to pick up another issue in the meantime if you feel like it :)

**Comment by ekpyron** [comment]
I'm sorry about not doing this sooner, but especially since we will replace the immutable logic by a code data location that will not require the write-once restriction in the constructor at all, the added complexity here is not worth the temporary benefits it may or may not bring.

So thank you very much for the contribution and for your efforts, and again I apologize for not clarifying this sooner, but I'll be closing this PR now.

**Pull request closed without merging**
